bip-txhash.md

```diff
@@ -0,0 +1,269 @@
+  cost is specified per TransactionHash execution. Using the same validation
+  budget ("sigops budget") introduced in BIP-0342, each TransactionHash
+  decreases the validation budget by 10. If this brings the budget below zero,
+  the script fails immediately.<br>The following considerations should be made:
+  * All fields that can be of arbitrary size are cachable as TransactionHash always hashes their hashed values.
+  * In "individual mode", a user can at most commit 32 inputs or outputs,
+    which we don't consider excessive for potential repeated use.
+  * In "leading mode", a caching strategy can be used where the SHA256 context
+    is stored every N in/outputs so that multiple executions of the
+    TransactionHash function can use the caches and only have to hash an
+    additional N-1 items at most.
+
+
+# Motivation
+
+This BIP specifies a basic transaction introspection primitive that is useful
+to either reduce interactivity in multi-user protocols or to enforce some basic
+constraints on transactions.
+
+Additionally, the constructions specified in this BIP can lay the groundwork for
+some potential future upgrades:
+* The TxFieldSelector construction would work well with a hypothetical opcode
+  `OP_TX` that allows for directly introspecting the transaction by putting the
+  fields selected on the stack instead of hashing them together.
+* The TransactionHash obtained by `OP_TXHASH` can be combined with a
+  hypothetical opcode `OP_CHECKSIGFROMSTACK` to effectively create an
+  incredibly flexible signature hash, which would enable constructions like
+  `SIGHASH_ANYPREVOUT`.
+
+## Comparing with some alternative proposals
+
+* This proposal strictly generalizes BIP-119's `OP_CHECKTEMPLATEVERIFY`, as the
+  default mode of our TxFieldSelector is effectively the same (though not
+  byte-for-byte identical) as what `OP_CTV` acomplishes, without costing any
+  additional bytes. Additionally, using `OP_CHECKTXHASHVERIFY` allows for more
+  flexibility which can help in the case for
+  * enabling adding fees to a transaction without breaking a multi-tx protocol;
+  * multi-user protocols where users are only concerned about their own inputs and outputs.
+
+* Constructions like `OP_IN_OUT_VALUE` used with `OP_EQUALVERIFY` can be
+  emulated by two `OP_TXHASH` instances by using the TxFieldSelector to select
+  a single input value first and a single output value second and enforcing
+  equality on the hashes. Neither of these alternatives can be used to enforce
+  small value differencials without the availability of 64-bit arithmetic in
+  Script.
+
+* Like mentioned above, `SIGHASH_ANYPREVOUT` can be emulated using `OP_TXHASH`
+  when combined with `OP_CHECKSIGFROMSTACK`: 
+  `<txfs> OP_TXHASH <pubkey> OP_CHECKSIGFROMSTACK` effectively emulates `SIGHASH_ANYPREVOUT`.
+
+
+# Detailed Specification
+
+A reference implementation in Rust is provided attached as part of this BIP
```

Comment:
OK, I would *really* appreciate a table of all the bits and exactly what and how they encode them.  It's particularly nasty because some values are little-endian 32 bit encoded, not CScriptNum encoded, and others are varint encoded?  

But it's nice to be explicit in each case, for people like me who are not deep in the weeds of bitcoin's onchain representation, since it helps when considering how to use this alongside things like OP_CAT and extended arithmetic opcodes.

Reply:
Ok, I agree. I think I tried to encode values the way they are consistently encoded in other contexts like sighashes and p2p. But I will go over them and list them in the BIP as well.